File: dynamicprogramming/camhoa/camhoa.py
import os
import sys

input_file = os.path.join(sys.path[0], 'camhoa.inp')
output_file = os.path.join(sys.path[0], 'camhoa.out')

k = 0  # số bó hoa
n = 0  # số lọ
values = []

# D[i][j] là tổng giá trị thẩm mỹ lớn nhất khi cắm i bó hoa đầu tiên vào j lọ hoa đầu tiên
# i: inflorescence (cụm hoa), j: jar (lọ)
D = []


def input():
    global k, n, values
    with open(input_file, 'r') as f:
        k, n = map(int, f.readline().split())

        values = [[0] * (n + 1) for _ in range(k + 1)]

        for i in range(1, k + 1):
            values[i][1:] = list(map(int, f.readline().split()))


def process():
    global k, n, values, D

    # Khởi tạo giá trị 0 cho toàn bảng quy hoạch D
    D = [[0 for j in range(n + 1)] for i in range(k + 1)]

    # Duyệt từng hàng i (inflorescence)
    for i in range(1, k + 1):
        # Duyệt từng cột j (jar)
        for j in range(1, n + 1):
            if i <= j:
                # TH1: lọ j không được chọn cắm
                j_not_chosen = D[i][j - 1]

                # TH2: lọ j được chọn cắm
                j_chosen = D[i - 1][j - 1] + values[i][j]

                # Lấy giá trị thẩm mỹ tốt nhất
                D[i][j] = max(j_not_chosen, j_chosen)


def output():
    global k, n, D

    # inflorescence
    # inflor[i] = j: chọn lọ j cho bó hoa i
    inflor = [0] * (k + 1)

    i = k
    j = n

    while i > 0 and j > 0:
        if i <= j and D[i][j] != D[i][j - 1]:
            # Chọn lọ j cho bó hoa i
            inflor[i] = j

            # Lùi về bó hoa liền trước
            i -= 1

            # Lùi về lọ liền trước
            j -= 1
        else:
            # Chỉ lùi về lọ liền trước
            j -= 1

    with open(output_file, 'w') as f:
        f.write(f'{D[k][n]}\n')

        for i in range(1, k + 1):
            f.write(f'{inflor[i]} ')


def show():
    for i in range(1, k + 1):
        for j in range(1, n + 1):
            print(f'{D[i][j]:>4}', end='')

        print()


if __name__ == '__main__':
    input()
    process()
    output()
    show()

File: dynamicprogramming/camhoa/camhoa.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <iomanip>

#define input_file "camhoa.inp"
#define output_file "camhoa.out"

using namespace std;

int k, n;
vector<vector<int>> values;

// D[i][j] là tổng giá trị thẩm mỹ lớn nhất khi cắm i bó hoa đầu tiên vào j lọ hoa đầu tiên
// i: inflorescence (cụm hoa), j: jar (lọ)
vector<vector<int>> D;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> k >> n;

    values.resize(k + 1, vector<int>(n + 1, 0));

    for (int i = 1; i < k + 1; ++i)
    {
        for (int j = 1; j < n + 1; ++j)
        {
            cin >> values[i][j];
        }
    }
}


void process()
{
    // Khởi tạo giá trị 0 cho toàn bảng quy hoạch D
    D.resize(k + 1, vector<int>(n + 1, 0));

    // giá trị thẩm mỹ lớn nhất trong trường hợp 1 và 2
    int j_not_chosen, j_chosen;

    // Duyệt từng hàng i (inflorescence)
    for (int i = 1; i < k + 1; ++i)
    {
        // Duyệt từng cột j (jar)
        for (int j = 1; j < n + 1; ++j)
        {
            // Nếu số bó hoa ít hơn hoặc bằng số lọ
            if (i <= j)
            {
                // TH1: lọ j không được chọn cắm
                j_not_chosen = D[i][j - 1];

                // TH2: lọ j được chọn cắm
                j_chosen = D[i - 1][j - 1] + values[i][j];
                
                // Lấy giá trị thẩm mỹ tốt nhất
                D[i][j] = max(j_not_chosen, j_chosen);
            }
        }
    }
}


void output()
{
    // inflorescence
    // inflor[i] = j: chọn lọ j cho bó hoa i
    vector<int> inflor(k + 1, 0);

    int i = k;
    int j = n;

    while (i > 0 && j > 0)
    {
        if (i <= j)
            if (D[i][j] != D[i][j - 1])
            {
                // Chọn lọ j cho bó hoa i
                inflor[i] = j;
                
                // Lùi về bó hoa liền trước
                --i;

                // Lùi về lọ liền trước
                --j;
            }        
        else
        {
            // Chỉ lùi về lọ liền trước
            --j;
        }
    }

    freopen(output_file, "w", stdout);

    cout << D[k][n] << '\n';

    for (int i = 1; i < k + 1; ++i)
    {
        cout << inflor[i] << ' ';
    }
}


void show()
{
    for (int i = 1; i < k + 1; ++i)
    {
        for (int j = 1; j < n + 1; ++j)
        {
            cout << setw(4) << D[i][j] << ' ';
        }

        cout << endl;
    }
}


int main()
{
    input();
    process();
    show();
    output();

    return 0;
}